src/compiler/parser.ts

```diff
@@ -3981,23 +3981,57 @@ module ts {
                     return parseDebuggerStatement();
                 case SyntaxKind.AtToken:
                     return parseDeclaration();
+
                 case SyntaxKind.ConstKeyword:
-                case SyntaxKind.DeclareKeyword:
                 case SyntaxKind.EnumKeyword:
                 case SyntaxKind.ExportKeyword:
                 case SyntaxKind.ImportKeyword:
-                case SyntaxKind.InterfaceKeyword:
-                case SyntaxKind.ModuleKeyword:
-                case SyntaxKind.NamespaceKeyword:
                 case SyntaxKind.PrivateKeyword:
                 case SyntaxKind.ProtectedKeyword:
                 case SyntaxKind.PublicKeyword:
                 case SyntaxKind.StaticKeyword:
-                case SyntaxKind.TypeKeyword:
                     if (getDeclarationFlags() & flags) {
                         return parseDeclaration();
                     }
                     break;
+
+                // 'declare', 'module', 'namespace', 'interface'* and 'type' are all legal JavaScript identifiers when
+                // ASI takes effect. In such cases, we cannot parse out the "expected" declarations. For instance, while
+                //
+                //    namespace n
+                //
+                // can be none other than the beginning of a namespace declaration, JavaScript sees
+                //
+                //    namespace
+                //    n
+                //
+                // as the identifier 'namespace' on one line followed by the identifier 'n' on another.
+                // We need to look one token ahead to see if it permissible to try parsing a declaration.
+                //
+                // *Note*: 'interface' is actually a strict mode reserved word. So while
+                //
+                //   "use strict"
+                //   interface
+                //   I {}
+                //
+                // could be legal, it would add complexity for very little gain.
+                case SyntaxKind.DeclareKeyword:
+                    if (lookAhead(nextTokenIsIdentifierOrKeywordOnSameLine) && getDeclarationFlags() & flags) {
+                        return parseDeclaration();
+                    }
+                    break;
+                case SyntaxKind.ModuleKeyword:
+                    if (lookAhead(nextTokenIsIdentifierOrStringLiteralOnSameLine) && getDeclarationFlags() & flags) {
+                        return parseDeclaration();
+                    }
+                    break;
+                case SyntaxKind.InterfaceKeyword:
+                case SyntaxKind.NamespaceKeyword:
+                case SyntaxKind.TypeKeyword:
+                    if (lookAhead(nextTokenIsIdentifierOnSameLine) && getDeclarationFlags() & flags) {

```

Comment:
Do you still need these declaration flags checks? Are they redundant of the check you are adding?